backend/windmill-queue/src/jobs.rs

```diff
@@ -684,9 +690,113 @@ pub async fn pull<R: rsmq_async::RsmqConnection + Clone>(
         }
     }
 
+    let final_workspaces_filter = workspaces_filter.as_str();
+
     // let rs = rd_string(2);
     // let instant = Instant::now();
 
+    let mut script_path_blocklist: Vec<String> = Vec::new();
+    loop {
+        // at each iteration, we need to ignore the script paths that have already been
+        // pulled and for which the concurrency limit has been reached. Otherwise each
+        // single worker will loop indefinitely here until the concurrency limit is not
+        // reached anymore
+        let cloned_sript_path_blocklist = script_path_blocklist.clone();
+        let mut script_path_filter = String::new();
+        if !cloned_sript_path_blocklist.is_empty() {
+            script_path_filter.push_str(&format!(
+                " AND script_path NOT IN ({})",
+                cloned_sript_path_blocklist
+                    .into_iter()
+                    .map(|x| format!("'{x}'"))
+                    .collect::<Vec<String>>()
+                    .join(",")
+            ));
+        }
+        let job = pull_single_job_and_mark_as_running_no_concurrency_limit(db, final_workspaces_filter, script_path_filter.as_str(), rsmq.clone()).await?;
+
+        if job.is_none() {
+            return Ok(None);
+        }
+
+        // concurrency check. If more than X jobs for this path are already running, we re-queue and pull another job from the queue
+        let pulled_job = job.unwrap();
+        if !pulled_job.script_path.is_none() {
+            let job_script_path = pulled_job.script_path.clone().unwrap();
+            script_path_blocklist.push(job_script_path.clone());
+
+            let job_custom_concurrent_limit = pulled_job.concurrent_limit.unwrap_or(DEFAULT_JOB_CONCURRENT_LIMIT);
+            let job_custom_concurrency_time_window_s = pulled_job.concurrency_time_window_s.unwrap_or(DEFAULT_JOB_CONCURRENCY_TIME_WINDOW_SECS);
+            
+            let concurrent_jobs_for_this_script = sqlx::query_scalar!(
+                "SELECT COALESCE(completed_count, 0) + COALESCE(running_count, 0) AS total_count
+                FROM
+                    (SELECT script_path, COUNT(*) as completed_count
+                    FROM completed_job
+                    WHERE script_path = $1 AND started_at > (now() - INTERVAL '1 second' * $2)
```

Comment:
you actually do not need to check for time, just that running = true :)

Reply:
it should simplify a lot the logic for the queue, it's just a count where script_path = x